pennylane/_device.py

```diff
@@ -216,12 +216,22 @@ def execute(self, queue, observables):
                     results.append(self.expval(obs.name, obs.wires, obs.parameters))
                 elif obs.return_type == "variance":
                     results.append(self.var(obs.name, obs.wires, obs.parameters))
+                elif obs.return_type == "sample":
+                    if not hasattr(obs, 'num_samples'):
+                        raise DeviceError("Number of samples not specified for observable {}".format(obs.name))
+
+                    results.append(self.sample(obs.name, obs.wires, obs.parameters, obs.num_samples))
 
             self.post_measure()
 
             self._op_queue = None
             self._obs_queue = None
 
+            # Ensures that a combination with sample does not put
+            # expvals and vars in superfluous arrays
+            if any(obs.return_type == "sample" for obs in observables):
```

Comment:
If your circuit returned
```python
qml.expval(qml.PauliX(0)), qml.var(PauliX(1)), qml.sample(qml.PauliX(2), 7)
```
then _without_ this check you get
```python
[[0.12], [0.57], [1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0]]
```
_with_ this check you get
```python
[0.12, 0.57, [1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0]]
```
which serves the intuition that circuit(x)[0] should be a float and not an array.

Reply:
Good that you pointed me to this -- I actually had an untested bug in there.

There are two layers of conversions that have been going on, first `Device.execute` converts the list of values it calculated to an numpy array. I updated this so that the output already has the proper format. I also kept the `np.asarray` in the QNode output conversion to be on the safe side if someone decides to overwrite `execute` in their device implementation.